instrumentation/jdbi3/src/main/java/brave/jdbi3/BraveStatementContextListener.java

```diff
@@ -0,0 +1,142 @@
+    private static final Pattern serviceNamePattern = Pattern.compile("(^|&)zipkinServiceName=(?<serviceName>[^&]*)");
+    private Map<String, String> serviceNameCache = new ConcurrentHashMap<String, String>() { };
+
+    public BraveStatementContextListener(Tracing tracing) { this.tracing = tracing; }
+
+    @Override
+    public void contextCreated(final StatementContext ctx) {
+        String spanName = ctx.describeJdbiStatementType();
+        Span span = tracing.tracer().nextSpan().name(spanName);
+
+        Instant start = ctx.getExecutionMoment();
+        if (start != null) {
+            span.start(start.toEpochMilli() * 1000);
+        } else {
+            span.start();
+        }
+
+        ctx.setTraceId(span.context().traceIdString());
+
+        ctx.addCleanable(() -> {
+            final SqlStatements stmtConfig = ctx.getConfig(SqlStatements.class);
+
+            String remoteServiceName = getRemoteServiceName(ctx);
+            if (remoteServiceName != null && !remoteServiceName.isEmpty()) {
+                span.remoteServiceName(remoteServiceName);
+            }
+
+            span.kind(Span.Kind.CLIENT);
+
+            final String renderedSql = ctx.getRenderedSql();
+            if (renderedSql != null) {
+                String truncated = renderedSql.substring(
+                    0,
+                    Math.min(renderedSql.length(), stmtConfig.getJfrSqlMaxLength())
+                );
+                span.tag("sql.query", truncated);
+            }
+
+            span.tag("sql.rows", Long.toString(ctx.getMappedRows()));
+
+            if (ctx.getCompletionMoment() == null) {
+                span.tag("error", "");
+            }
+
+            if (ctx.getCompletionMoment() != null) {
+                span.finish(ctx.getCompletionMoment().toEpochMilli() * 1000);
+            } else if (ctx.getExceptionMoment() != null) {
+                span.finish(ctx.getExceptionMoment().toEpochMilli() * 1000);
+            } else {
+                span.finish();
+            }
+        });
+    }
+
+    String getRemoteServiceName(StatementContext ctx) {
+        Connection connection = ctx.getConnection();
+        if (connection == null) {
+            return null;
+        }
+        URI uri;
+        try {
+            String url = connection.getMetaData().getURL();
+            if (serviceNameCache.containsKey(url)) {
+                return serviceNameCache.get(url);
+            }
+            if (url.startsWith("jdbc:")) {
+                // strip "jdbc:" prefix
+                url = url.substring(5);
+            }
+            uri = new URI(url); // strip "jdbc:"
+
+            String remoteServiceName = extractRemoteServiceName(uri);
+            serviceNameCache.put(url, remoteServiceName);
+            return remoteServiceName;
+        } catch (SQLException | URISyntaxException ignored) {
+            return null;
+        }
+    }
+
+    static String extractRemoteServiceName(URI uri) {
```

Comment:
Parsing JDBC URLs seems to be a large chunk of what this class is doing, probably worth extracting a separate `JdbcUrlParser `(so it could be also tested on different URLs).

Reply:
There are already plenty of URL parsing tests in `shouldParseRemoteSeriveName` (which is now part of BraveStatementContextListenerTest). What other tests do you have in mind?

Reply:
Yeah, I think my point is: `BraveStatementContextListenerTest::shouldParseRemoteSeriveName` does not test anything in `BraveStatementContextListener` besides JDBC URL parsing (for service name extraction) but needs like 5-6 mocks. The suggestion to have a dedicated parser would:
 - simplify `BraveStatementContextListener` since the large chunk of supporting code will go away
 - simplify testing by removing the need for mocking

What do you think?